### PR TITLE
Optimize Hydra Eval (rebased)

### PR DESCRIPTION
**Pull request by jbgi**
this is a rebased version of #2294 with most @rvl review comments implemented, except still using `release-lib`.

Evaluation times:
* [Evaluation PR #2315](https://hydra.iohk.io/jobset/Cardano/cardano-node-pr-2315/latest-eval) - 331s
* [Evaluation PR #2294](https://hydra.iohk.io/jobset/Cardano/cardano-node-pr-2294/latest-eval) - 328s
* [Evaluation master branch](https://hydra.iohk.io/jobset/Cardano/cardano-node/latest-eval) - 1513s

### CODE REVIEW

**Comment by newhoggy** [comment]
How much do these optimisations reduce eval time (or some other metric)?

**Comment by rvl** [comment]
@newhoggy I updated the PR description with eval times.

File: release.nix [hunk 1]
```diff
@@ -125,47 +141,66 @@ let
...
       inherit (pkgs.haskell-nix) hackageSrc stackageSrc;
     };
+  } // (optionalAttrs linuxBuild (with linuxRelease; {
+    linux =
+      let filteredBuilds = mapAttrsRecursiveCond (a: !(isList a)) (path: value:
+        if (any (p: take (length p) path == p) onlyBuildOnDefaultSystem) then filter (s: !(elem s nonDefaultBuildSystems)) value else value)
+        (packagePlatforms project);
+      in (mapTestOn (__trace (__toJSON filteredBuilds) filteredBuilds)) // {
+        # linux static builds:
+        musl64 = mapTestOnCross musl64 (packagePlatformsCross (filterProject noMusl64Build project));
```
Comment (rvl) [inline release.nix:156]: Can musl64 and native attrsets be on the same level?

File: release.nix [hunk 2]
```diff
@@ -125,47 +141,66 @@ let
...
-    cardano-node-linux = import ./nix/binary-release.nix {
-      inherit pkgs project;
-      platform = "linux";
-      exes = filter (p: p.system == "x86_64-linux") (collectJobs jobs.musl64.exes);
+      exes = collectJobs jobs.macos.exes;
     };
-  } // (optionalAttrs windowsBuild {
-    "${mingwW64.config}" = mapTestOnCross mingwW64 (packagePlatformsCross (filterProject noCrossBuild));
+  })) // (optionalAttrs windowsBuild (with windowsRelease; {
+    "${mingwW64.config}" = mapTestOnCross mingwW64 (packagePlatformsCross (filterProject noCrossBuild project));
```
Comment (rvl) [inline release.nix:175]: Rather than calling this attrset `x86_64-w64-mingw32`, how about just `windows`?

**Review by rvl — APPROVED** [review]
Looks good to me, thanks @hamishmack and @jbgi.

I am very curious to find out how much cardano-wallet eval time can be improved by the same hack.

One thing about the jobset. It was like this before, but still, do you really need to recurse into haskellPackages? It would probably be enough to have `exes.**`, `checks.**`, `benchmarks.**` in the jobset.

**Comment by angerman** [comment]
Can we get a write up on this? Retro/Lessons learned?

**Pull request merged**